Started making the reponse PSR7 compatible

burzum opened the pull request:
Replacing #9583 with a clean start.

Commit:
Started implementing the PSR7 response interface

Commit:
Coding standard fixes.

Commit:
Adding the PSR7 interface to the Response object

markstory commented on src/Network/Response.php:
```diff
@@ -640,10 +645,17 @@ public function header($header = null, $value = null)
         }
 
-        return $this->_headers;
+        return $this->headers;
```
Won't this change the return value of `header()` with no arguments as the internal storage is different?

burzum replied:
Hmm the tests didn't fail so... But I'll take a closer look.

markstory commented on src/Network/Response.php:
```diff
@@ -658,16 +670,33 @@ public function header($header = null, $value = null)
         if ($url === null) {
-            $headers = $this->header();
+            $result = $this->getHeaderLine('Location');
+            if (empty($result)) {
```
You could do `!$result` here.

markstory commented on src/Network/Response.php:
```diff
@@ -1072,10 +1163,10 @@ public function modified($time = null)
-        if (isset($this->_headers['Last-Modified'])) {
-            return $this->_headers['Last-Modified'];
+        if (isset($this->headers['Last-Modified'])) {
+            return $this->headers['Last-Modified'];
```
Should you be using hasHeader() and getHeaderLine() here?

markstory commented on src/Network/Response.php:
```diff
@@ -1072,10 +1163,10 @@ public function modified($time = null)
         if ($time !== null) {
             $date = $this->_getUTCDate($time);
-            $this->_headers['Last-Modified'] = $date->format('D, j M Y H:i:s') . ' GMT';
+            $this->headers['Last-Modified'] = $date->format('D, j M Y H:i:s') . ' GMT';
```
`_setHeader()` ?

markstory commented on tests/TestCase/Network/ResponseTest.php:
```diff
@@ -2096,4 +2096,65 @@ public function testLocation()
+        $response = $response->withStatus(404);
+        $reasonPhrase = $response->getReasonPhrase();
+        $this->assertEquals('Not Found', $reasonPhrase);
+    }
```
We'll need tests for the other header methods. I know they are implemented via a trait that also has tests, but if we ever change that it would be good to have tests covering the behavior we promised people.

burzum replied:
Done.

markstory reviewed (commented):
Looking good.

Commit:
Added the stream object to the Response class

Commit:
phpcs fix

Commit:
Refactoring the stream creation in Response.php

Commit:
Fixing a bug in the stream (re-)creation

burzum commented:
Just a side note: I'll be out for a week until I can continue with this.

Commit:
Adding tests for the new response class methods.

Commit:
Adding more tests for the response class.

burzum commented:
@markstory I'm back and added more tests, please gimme an updated on this as soon as you find the time.

markstory commented on src/Network/Response.php:
```diff
@@ -417,6 +430,20 @@ class Response
+            $this->_streamTarget = $options['streamTarget'];
+        }
+        if (isset($options['streamMode'])) {
+            $this->_streamMode = $options['streamMode'];
```
What is the purpose of these options? Do you have any scenarios where you would want to create a `r` stream in a response?

burzum replied:
No idea when *I* want to make it read only but there might be the cause *somebody* wants. I just didn't looked right to me that you can pass a stream but not a the options for it. I remove it if you want but I'm pretty sure we'll see a future ticket about this being added back.

markstory commented on src/Network/Response.php:
```diff
@@ -417,6 +430,20 @@ class Response
+        }
+        if (isset($options['stream'])) {
+            if (!$options['stream'] instanceof StreamInterface) {
+                throw new InvalidArgumentException('Stream option must be an object implement StreamInterface');
```
'must be an object that implements'

markstory commented on src/Network/Response.php:
```diff
@@ -677,10 +739,50 @@ public function location($url = null)
-        return $this->_body = $content;
+        // BC compatibility
+        if (is_callable($content)) {
+            $content = $this->_handleCallableBody($content);
```
Couldn't you use `CallbackStream` here? Right now you'll eagerly allocate all the memory needed to handle the response.

burzum replied:
Callbackstream - which one? Cake or Zend? Both use a different interface than the current Stream object plus you would create a new instance of the stream object here which would break the immutability of the original response?

markstory replied:
The CakePHP one is backwards compatible with how body() works now (no string is expected to be returned out of the callback)

The current `body` method can't be expected to work in an immutable context as it has to be backwards compatible. I think once this set of changes merge, we'll do a round of deprecations on methods that overlap the PSR7 interface and add in any missing get/with methods.

markstory commented on src/Network/Response.php:
```diff
@@ -677,10 +739,50 @@ public function location($url = null)
+        $this->stream->rewind();
+        $result = $this->stream->getContents();
+
+        if (empty($result)) {
```
Won't this break responses of `'0`'?

Commit:
Fixing return of values seen as empty in Response.php

markstory committed:
Don't eagerly call callback responses.

Don't eagerly evaluate callback/streaming responses. This can result in
allocating a pile of memory too soon in the request. Instead it is
preferrable to allocate that memory when we're flushing to the SAPI.

markstory commented:
Merging to unblock other work in #9636
